Improvements on OpenAPI spec generation

mbohlool opened the pull request:
- Generating models using go2idl library (no reflection anymore)
- Remove dependencies on go-restful/swagger
- Generate one swagger.json file for each web-service
- Bugfix: fixed a bug in trie implementation

Reference: #13414

**Release note**:

``` release-note
Generate separate OpenAPI spec for each API GroupVersion on /<Group>/<Version>/swagger.json
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31468)

<!-- Reviewable:end -->

mbohlool committed:
Add +k8s:openapi-gen tag to API types

mbohlool committed:
Update generated proto and swagger docs

thockin commented on cmd/libs/go2idl/openapi-gen/main.go:
```diff
@@ -0,0 +1,72 @@
+limitations under the License.
+*/
+
+// deepcopy-gen is a tool for auto-generating DeepCopy functions.
```
comments are about deepcopy

thockin commented on Makefile.generated_files:
```diff
@@ -286,6 +286,111 @@ $(DEEPCOPY_GEN):
+#
+# Any package that wants open-api functions generated must include a
+# comment-tag in column 0 of one file of the form:
+#     // +openapi=<VALUE>
```
so if this is +k8s:openapi then the `find` below is faster.  It's a minor point, but it is consistent with other "updated" tags we're using.  If/when we bust this out of the main repo, it's something we'd have to reconsider, but for now I suggest it.

thockin replied:
also, the other tags use the binary name as the tag, so "openapi-gen" rather than "openapi"

thockin commented on Makefile.generated_files:
```diff
@@ -286,6 +286,111 @@ $(DEEPCOPY_GEN):
+ifeq ($(DBG_MAKEFILE),1)
+    $(warning ***** finding all +openapi tags)
+endif
+OPENAPI_DIRS := $(shell                             \
```
nit. but I tried to have all the \ end-of-line align :)

thockin commented on cmd/libs/go2idl/openapi-gen/README:
```diff
@@ -0,0 +1,10 @@
+$VALUE can be any of these:
+
+- "package": Generate spec for all types in this package (should only be added to package comments)
+- "definition": Generate spec for a single type (should only be added to type comments)
```
optional: the convention used in deepcopy and conversion is that a "package" tag can be overriden at the type, and the values are true or false.  I would be fine to see "true, false, should-fail" ?

thockin commented:
I *only* reviewed the scaffolding for this, not the actual logic.  As discussed, I think we should understand any slowdown this imposes on builds and decide if it is really worthwhile.

mbohlool committed:
Don't remvove generated files in staging folder

mbohlool committed:
Add Finalize method to go2idl Generator interface

mbohlool committed:
Customize OpenAPI definition types

customize intstr, unversioned.time, and quantity OpenAPI definitions.

mbohlool committed:
Improvements on OpenAPI spec generation:

- Generating models using go2idl library (no reflection anymore)
- Remove dependencies on go-restful/swagger
- Generate one swagger.json file for each web-service
- Bugfix: fixed a bug in trie implementation

mbohlool commented:
I've changed generated files to one file. This will decrease compile time overhead from 7% to 3.5% that is acceptable. This is ready for review. Please take another look.

thockin commented:
Makefile LGTM

lavalamp commented on pkg/apis/apps/doc.go:
```diff
@@ -15,5 +15,6 @@ limitations under the License.
 */
 
 // +k8s:deepcopy-gen=package,register
+// +k8s:openapi-gen=true
```
Does this apply to the whole package when it appears here?

mbohlool replied:
Yes, I had =package before but saw it redundant, the README file briefly cover this.

lavalamp commented on Makefile:
```diff
@@ -208,7 +208,7 @@ clean_meta:
 .PHONY: clean_generated
 clean_generated:
-	find . -type f -name $(GENERATED_FILE_PREFIX)\* | xargs rm -f
+	find . -type f -name $(GENERATED_FILE_PREFIX)\* | grep -v "./staging/*" | xargs rm -f
```
did you mean `"[.]/staging/.*`?

lavalamp commented on pkg/api/unversioned/time.go:
```diff
@@ -20,6 +20,9 @@ import (
 	"encoding/json"
 	"time"
 
+	"k8s.io/kubernetes/cmd/libs/go2idl/openapi-gen/generators"
```
I think we should not have a dependency on the generator package. Can we move the OpenAPIDefinition type into a 3rd package that both the generator and this can depend on? It's fine if it lives as a subdirectory in the generator. I just don't want to pull in all of the generator's dependencies here.

lavalamp commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,544 @@
+
+// OpenAPIGetDefinition describe types that can customize openAPI definitions. If a type implements this interface, the definition returns
+// by it will be used, otherwise the auto-generated definitions will be used.
+type OpenAPIGetDefinition interface {
```
I think the go style is something like OpenAPIDefiner or OpenAPIDefinitionGetter.

Like I said in another comment, I think these first three types should go in another package.

lavalamp commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,544 @@
+		glog.V(5).Infof("considering pkg %q", i)
+		pkg := context.Universe[i]
+		if pkg == nil {
+			// If the input had no Go files, for example.
```
tiny nit: might be better to do
```
pkg, ok := ...
if !ok { continue }
```

mbohlool replied:
that was a copy-paste left-over. sorry for that :)

lavalamp commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,544 @@
+		}
+		for _, t := range pkg.Types {
+			if hasOpenAPITagValue(t.CommentLines, tagTargetType) {
+				glog.V(5).Infof("target type : %q", t)
```
target means it's the destination package?

mbohlool replied:
yes, do you suggest we change the name to destination?

lavalamp commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,544 @@
+// openApiGen produces a file with auto-generated OpenAPI functions.
+type openAPIGen struct {
+	generator.DefaultGen
+	targetPackage string
```
maybe document this means.

mbohlool replied:
targetpackage was redundant, I've removed it and commented on targetType.

lavalamp commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,544 @@
+func (g *openAPIGen) Imports(c *generator.Context) []string {
+	importLines := []string{}
+	for _, singleImport := range g.imports.ImportLines() {
+		if g.isOtherPackage(singleImport) {
```
Is this still necessary if you grab https://github.com/kubernetes/gengo/pull/6? Definitely the import tracker should be doing this for you.

mbohlool replied:
I've added the change here. Is there any standard way to pull commits that are not merged yet into my PR? mine could (theoretically :) ) merge before that.

lavalamp commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,544 @@
+func isPropertyRequired(m *types.Member) bool {
+	jsonTags := getJsonTags(m)
+	// A property is required if it does not have omitempty value in its json tag
+	return len(jsonTags) < 2 || jsonTags[1] != "omitempty"
```
Does omitempty have to occur at location 1? I think there may be a few options that could appear in any order.

mbohlool replied:
all examples I could find, they put it in second location (even go documentation has a leading comma for omitempty when it is alone and put a note on that), but implementation of json package does not depend on it to be on second location. I will change this.

mbohlool replied:
ok, scratch that, it should be at least in the second position according to implementation.

lavalamp commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,544 @@
+		}
+		r := mt.Signature.Results[0]
+		if r.Name.Name != "OpenAPIDefinition" || r.Name.Package != openAPIGenPackagePath {
+			glog.Errorf("Err4 %v", t)
```
debug line? :)

mbohlool replied:
Oops.

lavalamp commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,544 @@
+	switch t.Kind {
+	case types.Struct:
+		args := argsFromType(t)
+		g.Do("\"$.$\": ", filepath.Base(t.Name.Package)+"."+t.Name.Name)
```
Does this do the same thing? This is what the name system is for.

```go
g.Do(`"$.|raw$": `, t)
```

lavalamp replied:
Or wait. I see. You're constructing a string name, not a go type name.

lavalamp replied:
So I guess this is fine. I guess you could turn that string concatenation into a function if you do it in more than one place, to be self documenting. Only if you feel like it.

mbohlool replied:
factored it out into another method.

lavalamp commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,544 @@
+	if t.Kind == types.Pointer {
+		t = t.Elem
+	}
+	t = resolvedAliasType(t)
```
Two is a suspicious number of times to call a function. Should this be done in a loop? :)

mbohlool replied:
Done.

lavalamp commented on pkg/genericapiserver/openapi/openapi.go:
```diff
@@ -16,213 +16,138 @@ limitations under the License.
-		return err
+func (o *openAPI) init() error {
+	if o.openAPIDefinitions == nil {
+		if openAPIGetType, ok := interface{}(o).(interface {
```
I guess this is so that you get a run-time error instead of a compile time error if you forget to run the generator? This is worth a comment, probably.

lavalamp replied:
Is a run time error really better?

lavalamp commented on pkg/genericapiserver/openapi/openapi.go:
```diff
@@ -16,213 +16,138 @@ limitations under the License.
+// by k8s.io/kubernetes/cmd/libs/go2idl/openapi-gen.
+func (o *openAPI) buildDefinitionForType(sample interface{}) (string, error) {
+	t := reflect.TypeOf(sample)
+	if t.Kind() == reflect.Ptr {
```
If this gets called once per type on startup then it doesn't matter, but if not then in a followup it might be reasonable to replace this with a generated type switch rather than using reflection. It probably won't get called often enough to matter. I just mention this because we've had to do similar things in the api encode/decode paths.

mbohlool replied:
this is only called once per type per swagger generation that is O(N) at startup (N is number of endpoints/~VersionKinds). I don't think this would be a (any) performance issue. I hope that sometime in future I can auto-generate routes/paths or at least improve it a little. In that change this would be among the firsts to go.

lavalamp commented on pkg/master/master_test.go:
```diff
@@ -1258,12 +1269,61 @@ func TestValidOpenAPISpec(t *testing.T) {
+			}
+			t.Logf("Validating open API spec on %v ...", path)
+
+			/*			resp, err = http.Get(server.URL + path + "/swagger.json")
```
delete dead code? (or resurrect?)

mbohlool replied:
This commented out code is very useful for debugging. Would you mind me keeping it, or do you have any suggestion for a place for it to live? I use this to generate spec files and put them in temp folder when I run the test. This is much much faster than bringing up an API server and hit its endpoints.

lavalamp commented:
I left comments on the individual commits. Nothing major, mostly looks awesome. I reviewed the generator pretty carefully and the openapi runtime part less carefully, maybe Nikhil can look at that.

mbohlool committed:
Only add to tracker if it's non-local

mbohlool commented:
Applied Daniel's comments. @nikhiljindal friendly ping.

mbohlool commented:
The test failures is because of new types added. I will rebase but it would be nice if we can get this in faster as it is a pain to rebase and build everything again.

nikhiljindal commented on Makefile.generated_files:
```diff
@@ -286,6 +286,107 @@ $(DEEPCOPY_GEN):
+# How to build the generator tool.  The deps for this are defined in
+# the $(OPENAPI_GEN).mk, above.
+#
+# A word on the need to touch: This rule might trigger if, for example, a
```
Thanks for detailed comments :0

nikhiljindal commented on cmd/libs/go2idl/openapi-gen/generators/common/common.go:
```diff
@@ -1 +1,65 @@
+type OpenAPIDefinitions map[string]OpenAPIDefinition
+
+// OpenAPIDefinitionGetter gets openAPI definitions for a given type. If a type implements this interface,
+// the definition returns by it will be used, otherwise the auto-generated definitions will be used.
```
returns -> returned?

nikhiljindal commented on cmd/libs/go2idl/openapi-gen/generators/common/common.go:
```diff
@@ -1 +1,65 @@
+}
+
+// This function is a reference for converting go (or any custom type) to a simple open API type,format pair
+func GetOpenAPITypeFormat(typeName string) (string, string) {
```
Add a comment regarding when should I add my custom type here and when should I add an OpenAPIDefinition() method. Maybe keep this method only for Go types and say that we need to add method for custom types?
For ex: for intOrString, I could have added it to this map instead of adding the OpenAPIDefinition() method as done in this PR, right?

mbohlool replied:
tried my best to explain in the comment, but simply, decision of not adding intOrString here was to keep the type and property documentation in the spec. if we add it here, the type documentation will be gone as we will have a simple type,format,doc (that doc is the property documentation).

nikhiljindal replied:
Good to know that. Hadnt realised it from reading the code.
Thanks for adding the comment.

nikhiljindal commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,521 @@
+)
+
+// openApiGen produces a file with auto-generated OpenAPI functions.
+type openAPIGen struct {
```
This implements Generator right?
At other places, we ensure and document that by adding the following line:
```
var _ Generator := openAPIGen{}
```

mbohlool replied:
I think you only need to do that if you are not converting the type to an interface (to make sure we will get compiler error). Here NewOpenAPIGen (the method immediately after it) is doing that, so I don't think this is necessary. If we miss one of the methods, we will get a compilation error.

nikhiljindal commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,521 @@
+			g.Do("Required: []string{\"$.$\"},\n", strings.Join(required, "\",\""))
+		}
+		g.Do("},\n},\n", nil)
+		// 				SchemaProps: spec.SchemaProps{
```
remove?

nikhiljindal commented on cmd/libs/go2idl/openapi-gen/generators/openapi.go:
```diff
@@ -0,0 +1,521 @@
+
+	// According to OpenAPI examples, only map from string is supported
+	if keyType.Name.Name != "string" {
+		return fmt.Errorf("Map with non-string keys are not supported by OpenAPI in %v", t)
```
We keep the first letter of error small, since this can be concatenated with other errors.

nikhiljindal commented on cmd/libs/go2idl/openapi-gen/main.go:
```diff
@@ -0,0 +1,44 @@
+limitations under the License.
+*/
+
+// This package generate openAPI definition file to be used in open API spec generation on API servers. To generate
```
generate -> generates

nikhiljindal commented:
pkg/genericapiserver/openapi/zz_generated.openapi.go is a huge file.
I see that for other generated files, we generate them one for each group version, for ex: zz_generated.deepcopy.go.
Is putting them all in one file a conscious decision? Do we plan to change that in future?

nikhiljindal commented on Makefile:
```diff
@@ -212,7 +212,7 @@ clean_meta:
 .PHONY: clean_generated
 clean_generated:
-	find . -type f -name $(GENERATED_FILE_PREFIX)\* | xargs rm -f
+	find . -type f -name $(GENERATED_FILE_PREFIX)\* | grep -v "[.]/staging/.*" | xargs rm -f
```
Why should we not remove files in staging folder? Will be good to add a comment here explaining that

mbohlool replied:
because generate does not generate them. Adding a comment can be useful, but the unwritten logic for clean_generated and generated_files is, whatever generated_files generates, clean_generated should clean.

nikhiljindal replied:
cc @caesarxuchao as FYI for staging specific code.

caesarxuchao replied:
It looks good.

nikhiljindal commented on cmd/libs/go2idl/generator/generator.go:
```diff
@@ -110,6 +110,10 @@ type Generator interface {
 	// Generators.)
 	Init(*Context, io.Writer) error
 
+	// Finalize should write finish up functions, and any other content that's not
```
Sorry, what do you mean by "finish up functions"? Probably giving an example will help?

mbohlool replied:
added more comment there. Basically it should be finish up code, the example is the exact thing I did here, generating one function for all types. Finalize used to close the function block. Added some comment to clarify it.

nikhiljindal commented on cmd/libs/go2idl/client-gen/test_apis/testgroup.k8s.io/v1/types.go:
```diff
@@ -25,8 +25,9 @@ import (
-	api.ObjectMeta       `json:"metadata,omitempty"`
-	Status               TestTypeStatus `json:"status,omitempty"`
+	// +k8s:openapi-gen=false
+	api.ObjectMeta `json:"metadata,omitempty"`
```
Why specially add a false here?

mbohlool replied:
this is only a test type and adding unversioned objectMeta bring in a whole set of unversioned objects in the generate file that is not used anywhere (and honestly it was a pain to pinpoint). I decided to just exclude it, because this is a test type.

nikhiljindal replied:
Will be good to add that comment in code.

nikhiljindal commented:
I am a bit confused by your commit order. Your commit adding the annotation and adding the generated file is before the commit that adds the generator :)

mbohlool commented:
Having a huge file was a conscious decision. It cut the compile time of the new code by half (from 7% added compile time to 3.5%) and also cut 25% of "make quick-release" time. For some reason quick-release does not like many added files. As it has benefit on compile time too, I've decided to put everything in one file. I think it is easier to manage to and as the file is not suppose to be edited by human, it does not matter how big it is.

mbohlool commented:
about the commit order, I don't think it does matter as long as they are in the same PR. Logical order would force to have the generator code as the first commit, or at least before any generated and tagging commits, but having generator commit at the end helps to rebase things easier. But I can reorder them as much as it does not make rebase harder.

mbohlool committed:
Generated open-api definition file

mbohlool commented:
I feel like a rebase master, so I reorder things logically :) Let see how it works. @nikhiljindal I think I applied or replied to all of your comments, please take another look. Thanks.

nikhiljindal commented:
re: commit order: I was just curious how you wrote code :) Did you write code in the order of commits or rearranged commits after that.

Just one optional comment to add a comment at another place.
Feel free to self apply the LGTM label.

Thanks for working on this and thanks for your patience :)

mbohlool commented:
Thanks @nikhiljindal @lavalamp @thockin

smarterclayton commented:
The generated file `pkg/genericapiserver/openapi/zz_generated.openapi.go` doesn't belong there.  It needs to be in a kubernetes specific package that is clearly marked.  We will have lots of generated API docs that are not part of the kube code base that have to be consistent.  I would recommend `pkg/api/openapi` or `pkg/openapi`.

smarterclayton commented on pkg/util/intstr/intstr.go:
```diff
@@ -101,6 +105,17 @@ func (intstr IntOrString) MarshalJSON() ([]byte, error) {
 	}
 }
 
+func (_ IntOrString) OpenAPIDefinition() common.OpenAPIDefinition {
```
This seems really wrong - now we're coupling every Go type that ever uses IntOrString with commonapi?  

This really needs to be outside this type as a separate package.

lavalamp replied:
I see your point, but at the same time intOrString is effectively part of our API. Maybe we should split it into two types, and move a copy into our api somewhere? (My preference is to nuke IntOrString from orbit but I think that's not on the table at the moment :) )

smarterclayton commented:
I have a couple of comments I'd like addressed in a follow up - the two points make it harder for extension APIs on top of Kube to integrate.
